### PR TITLE
Patch version update CI.

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Adds a new CI job to update the patch version of the helm charts. This job is trigger by an event send from other Kubewarden repositories (e.g. kubewarden-controller). This CI job will sync the CRDs when necessary as well as update the container image tag used. The container image tag is the version sent from the event payload sent by the other repositories.

Related to https://github.com/kubewarden/kubewarden-controller/issues/352
Fix https://github.com/kubewarden/helm-charts/issues/160
Fix https://github.com/kubewarden/helm-charts/issues/159

### CODE REVIEW

File: .github/workflows/update-charts.yml [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                    core.setOutput("update_type", "major")
+
+            } else if(oldVersion[1] != version[1]){
+                    core.setOutput("update_type", "minor")
+
+            } else if(oldVersion[2] != version[2]){
+                    core.setOutput("update_type", "patch")
+
+            } else {
+                    core.setFailed("Invalid version format")
```
Comment (flavio) [inline .github/workflows/update-charts.yml:34]: I think this happens if there's no difference between the old version and the new one

File: updatecli/updatecli.d/patch.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+targets:
+  chart-patch-version-update:
+    name: Bump chart patch version
+    kind: yaml
+    sourceid: chart
+    transformers:
+      - semverinc: "patch"
+    spec:
+      file: "file://charts/{{ requiredEnv .chart  }}/Chart.yaml"
+      key: "version"
```
Comment (flavio) [inline updatecli/updatecli.d/patch.yaml:21]: don't we have to bump also the `appVersion`?

There are also a ton of rancher specific variables that hold a reference to application version. These are the painful ones to update.
For example, take a look at this commit: https://github.com/kubewarden/helm-charts/commit/5d351b0d4ddb84b0789f19ead23d5d7a64cd54b1

File: .github/workflows/update-charts.yml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+      - name: Update CRDs Helm chart
+        if: steps.update_crds.outputs.must_update_crds_chart == 1
+        env:
+          UPDATECLI_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          UPDATECLI_CHART_NAME: "kubewarden-crds"
+        run: "updatecli apply --config ./updatecli/updatecli.d/patch.yaml --values updatecli/values.yaml"
+
+      - name: Commit changes
+        run: "updatecli/scripts/commit-changes.sh"
```
[the remark this answers is not in the document]
Comment (viccuad) [reply .github/workflows/update-charts.yml:125]: Agree here; I'm happy with a PR that bumps crds and controller image etc, but those PRs will not be intelligent enough to know when there's new functionality that needs to be exposed in values.yaml, or when they are introducing a breaking change.
So I think bumps to `version` and `appVersion` should happen manually still.

File: .github/workflows/update-charts.yml [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    steps:
+      - name: Check if it is patch update
+        id: step1
+        uses: actions/github-script@v6
+        with:
+          script: |
+            let oldVersion = context.payload.client_payload.oldVersion.split(".")
+            let version = context.payload.client_payload.version.split(".")
+            if (oldVersion.length != 3 && version.length != 3) {
+                    core.setFailed("Invalid version format")
```
Comment (viccuad) [inline .github/workflows/update-charts.yml:22]: goalpost moving, but I suppose this breaks for rcs.

File: .github/workflows/update-charts.yml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        env:
+          UPDATECLI_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: "updatecli apply --config ./updatecli/updatecli.d/patch.yaml --values updatecli/values.yaml"
+
+      - name: Update CRDs Helm chart
+        if: steps.update_crds.outputs.must_update_crds_chart == 1
+        env:
+          UPDATECLI_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          UPDATECLI_CHART_NAME: "kubewarden-crds"
+        run: "updatecli apply --config ./updatecli/updatecli.d/patch.yaml --values updatecli/values.yaml"
```
Comment (viccuad) [inline .github/workflows/update-charts.yml:122]: I wonder if we should not update the values.yaml of `kubewarden-crds` for now, as there's none.  That, or add an empty values.yaml so there's no change in this step?

File: updatecli/updatecli.d/patch.yaml (hunk 2 above)
Comment (viccuad) [reply updatecli/updatecli.d/patch.yaml:21]: > don't we have to bump also the `appVersion`?

Supposedly not, [by the rfc](https://github.com/kubewarden/rfc/blob/main/rfc/0007-kubewarden-versioning.md#a-patch-is-done-to-kubewarden-controller).

File: updatecli/scripts/commit-changes.sh [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+git config user.email "kubewarden@suse.de"
+git config user.name "Kubewarden bot"
+
+files=$(git --no-pager diff --name-only charts)
+echo $files
+echo "-----------------------------------"
+
+git commit -m "Update helm charts" \
+	-m "Commit changes made by the CI jobs to update the helm chart values and versions" \
```
[the remark this answers is not in the document]
Comment (viccuad) [reply updatecli/scripts/commit-changes.sh:11]: we could also form a url pointing to the specific workflow run, by using the [GITHUB_ env](https://docs.github.com/en/actions/learn-github-actions/environment-variables) vars :D. But maybe not worth the effort.

File: .github/workflows/update-charts.yml (hunk 1 above)
Comment (olblak) [reply .github/workflows/update-charts.yml:34]: I agree that semver can be more complicated that and sooner than later people will want to add metadata information.
Another library https://github.com/Masterminds/semver

File: .github/workflows/update-charts.yml (hunk 4 above)
Comment (flavio) [reply .github/workflows/update-charts.yml:22]: I think this should still work also for RCs. We are breaking up the version by `.`, hence `1.0.0-rc1` would be broken into `["1", "0", "0-rc1"]` which should be fine.

Maybe I didn't get it :sweat_smile:

File: .github/workflows/update-charts.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (olblak) [reply .github/workflows/update-charts.yml:125]: Also if the initial intend is commit without human validation, I would still open a pullrequest with the automerge flag enabled. 
Opening a pullrequest this way would have the following advanges

1. Visibility, as we would see the PR open and then merged in github.com/notifications
2. Test, It would trigger the test jobs which need to succeed before merging

File: .github/workflows/update-charts.yml (hunk 4 above)
Comment (viccuad) [reply .github/workflows/update-charts.yml:22]: no, you are right, it should still work :). Resolving.

File: updatecli/scripts/commit-changes.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (olblak) [reply updatecli/scripts/commit-changes.sh:11]: If you use the pullrequest approach with automerge and merge method squash then in the end you still have one commit

such as 

```
actions: # Action only work with Updatecli version > 0.40
  default:
    title: '[updatecli] Bump Epinio version used within installation documentation to {{ source "epinio" }}'
    kind: github/pullrequest
    scmid: default
    spec:
      automerge: true
      mergemethod: squash
      labels:
        - dependencies
```

File: .github/workflows/update-charts.yml (hunk 1 above)
Comment (jvanz) [reply .github/workflows/update-charts.yml:34]: Fair enough. I changed job to use the script shared by Flavio. But I would be great create a CLI using the Mastermind/semver.

File: .github/workflows/update-charts.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jvanz) [reply .github/workflows/update-charts.yml:125]: Okay, I'll open a PR then.

File: updatecli/updatecli.d/update-container-image-tag.yaml [hunk 7]
```diff
@@ -25,7 +26,38 @@ targets:
...
+    kind: "github/pullrequest"
+    scmid: "default"
+    spec:
+      automerge: true
+      mergemethod: squash
+      description: "Helm chart update"
+      draft: false
+      labels:
+        - "kind/enhancement"
+      title: "Helm chart update"
```
Comment (flavio) [inline updatecli/updatecli.d/update-container-image-tag.yaml:49]: I don't know if that helps, but updatecli can create a GH PR too. Maybe this would allow us to reduce the number of GH actions we depend upon

Comment (olblak) [reply updatecli/updatecli.d/update-container-image-tag.yaml:49]: @flavio That's what the "actions" do

On the updatecli side, we are currently deprecating `pullrequests` in favor of `actions`
In very very short, a pullrequest is github specific and we want to move the git logic for a target to a action.
More information on https://github.com/updatecli/updatecli/issues/465

You can run `updatecli manifest upgrade --config updatecli/updatecli.d` on your manifest to see what would change for your current updatecli version and `updatecli manifest upgrade --config updatecli/updatecli.d --in-place` to effectively upgrade your manifest if you are happy with the result.

File: updatecli/scripts/commit-changes.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (jvanz) [reply updatecli/scripts/commit-changes.sh]: Just awsering the Flavio's question. Yes, if we used the script to commit changes, the commit message will be as the first option. Git adds a empty line between the text defined in the `-m` command line flag.

[the remark this answers is not in the document]
Comment (jvanz) [reply updatecli/scripts/commit-changes.sh]: But as we will open a PR for the update. The script is not necessary.

File: .github/workflows/update-charts.yml (hunk 5 above)
Comment (jvanz) [reply .github/workflows/update-charts.yml:108]: Sorry, I do not understand the comment.

Comment (jvanz) [reply .github/workflows/update-charts.yml:108]: This is a typo. The last change does not have this change.

File: .github/workflows/update-charts.yml [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  check-update-type:
+    name: Detect update type
+    runs-on: ubuntu-latest
+    outputs:
+      update_type: ${{ steps.check_update_type.outputs.update_type }}
+    steps:
+      - name: Install semver comparison tool
+        run: |
+          wget -O ./semver https://github.com/fsaintjacques/semver-tool/raw/3.4.0/src/semver
+          chmod +x ./semver
```
Comment (flavio) [inline .github/workflows/update-charts.yml:17]: Nitpicking, you can download the binary to a dedicated dir and then "register" this dir into `$GITHUB_PATH` to make the command available (and save the `./semver` thing when invoking it).

Something like we're doing here:
https://github.com/kubewarden/github-actions/blob/18f8be2e8d95bfbb72333cad1bebbe48c9b06827/kwctl-installer/action.yaml#L16-L29

File: test [hunk 9]
```diff
@@ -0,0 +1 @@
+sfsdfd
```
Comment (flavio) [inline test:1]: Is this file needed? :thinking:

File: updatecli/updatecli.d/patch.yaml (hunk 2 above)
Comment (flavio) [reply updatecli/updatecli.d/patch.yaml:50]: @jvanz I think we have to update this updatecli "configuration" to keep the `version` and the `catalog.cattle.io/upstream-version` values in sync, see what we did manually inside of https://github.com/kubewarden/helm-charts/commit/5d351b0d4ddb84b0789f19ead23d5d7a64cd54b1

**Review by flavio — CHANGES_REQUESTED** [review]
Thanks for having addressed my concerns. I like this PR and I think we're close to being able to merge it.

The only blocker so far is the lack of automation around the `catalog.cattle.io/upstream-version` key (see one of the comments)

File: test (hunk 9 above)
Comment (jvanz) [reply test:1]: Ops... `git add .` bit me. hehe

File: .github/workflows/update-charts.yml [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        run: |
+          tar -xvf crds.tar.gz
+          find . -maxdepth 1 -name "*policyserver*" -exec  mv \{\} charts/kubewarden-crds/templates/policyserver.yaml \;
+          find . -maxdepth 1 -name "*admissionpolicies*" -exec  mv \{\} charts/kubewarden-crds/templates/admissionpolicy.yaml \;
+          find . -maxdepth 1 -name "*clusteradmissionpolicies*" -exec  mv \{\} charts/kubewarden-crds/templates/clusteradmissionpolicies.yaml \;
+
+          set +e
+          git diff --exit-code --no-patch charts/kubewarden-crds
+          echo "must_update_crds_chart=$?" >> $GITHUB_OUTPUT
+          git diff
```
Comment (jvanz) [inline .github/workflows/update-charts.yml:104]: @olblak how can I add these files in the branch used to open the PR? I can open a PR with the files changed using updatecli (yaml files). However, I could find a workaround to add files from external source in the commit

**Comment by jvanz** [comment]
@flavio, I've updated the script to update the annotations versions and fixing the issues that you reported in your last review. I just need to [talk with Olivier](https://github.com/kubewarden/helm-charts/pull/173#discussion_r1053707183) to find a solution to an issue after we decided to use PR to merge the changes.

File: .github/workflows/update-charts.yml (hunk 10 above)
Comment (olblak) [reply .github/workflows/update-charts.yml:104]: I need to try it but I would move 
```
          tar -xvf crds.tar.gz
          find . -maxdepth 1 -name "*policyserver*" -exec  mv \{\} charts/kubewarden-crds/templates/policyserver.yaml \;
          find . -maxdepth 1 -name "*admissionpolicies*" -exec  mv \{\} charts/kubewarden-crds/templates/admissionpolicy.yaml \;
          find . -maxdepth 1 -name "*clusteradmissionpolicies*" -exec  mv \{\} charts/kubewarden-crds/templates/clusteradmissionpolicies.yaml \;
```

to a shell script then call that shell script from the target of kind "shell"

**Comment by jvanz** [comment]
@flavio @olblak these are PRs created by the script from this PR:
Patch update with CRD update: https://github.com/jvanz/helm-charts/pull/31 
Patch update with no CRD update: https://github.com/jvanz/helm-charts/pull/32

File: updatecli/updatecli.d/patch-kubewarden-controller-with-crds-update.yaml [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        #    sourceid: admissionPolicySource
+        #    spec:
+        #      command: bash updatecli/scripts/install_crds.sh
+        #  installClusterAdmissionPolicyCRD:
+        #    name: "Install cluster admission policy CRD"
+        #    kind: shell
+        #    scmid: "default"
+        #    sourceid: clusterAdmissionPolicySource
+        #    spec:
+        #      command: bash updatecli/scripts/install_crds.sh
```
Comment (flavio) [inline updatecli/updatecli.d/patch-kubewarden-controller-with-crds-update.yaml:68]: nitpicking: do we need these commented lines?

File: updatecli/updatecli.d/patch-kubewarden-controller-with-crds-update.yaml [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+      #      command: echo policyserver
+      #  admissionPolicySource:
+      #    kind: shell
+      #    spec:
+      #      command: echo admissionpolicies
+      #  clusterAdmissionPolicySource:
+      #    kind: shell
+      #    spec:
+      #      command: echo clusteradmissionpolicies
+
```
Comment (flavio) [inline updatecli/updatecli.d/patch-kubewarden-controller-with-crds-update.yaml:39]: nitpicking: do we need these commented lines?

**Review by flavio — APPROVED** [review]
LGTM, I just left two minor comments

File: updatecli/updatecli.d/patch-kubewarden-controller-with-crds-update.yaml (hunk 11 above)
Comment (jvanz) [reply updatecli/updatecli.d/patch-kubewarden-controller-with-crds-update.yaml:68]: Sorry. I forgot to remote them. Fixed it.

File: updatecli/updatecli.d/patch-kubewarden-controller-with-crds-update.yaml (hunk 12 above)
Comment (jvanz) [reply updatecli/updatecli.d/patch-kubewarden-controller-with-crds-update.yaml:39]: Sorry. I forgot to remote them. Fixed it.
